[anyrun] Migrate feed and task connectors to anyrun-sdk and add a TI Lookup connector

semayellow opened the pull request:
<!--
Thank you very much for your pull request to the OpenCTI project! We as a community driven project depend on support and contributions like this!

Thus already a BIG THANK YOU upfront to you for choosing to help with your PR.
-->

### Proposed changes

* connector-anyrun-task
  * improve and extend documentation and configuration files
  * refactor and decompose the logic of the main module
  * improve indicators processing
  * add analysis report loading
  * temporary remove mitre, process features (due to internal technical problems)
* connector-anyrun-feed
  * improve and extend documentation and configuration files
  * refactor the logic of the main module
* connector-anyrun-lookup (new)

### Related issues

No related issues

### Checklist

<!--
Please submit the source code in a way, where you could honestly say `This code is finished`.
If you feel that there are possibilities for improving the code quality, please do so.
By doing this, you are actively helping us to improve the quality of the entire OpenCTI project.
-->

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

<!-- For completed items, change [ ] to [x]. -->
<!-- To sign commits with a GPG key, you can refer to https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits. -->

### Further comments

This PR aims to improve further connectors maintenance using [ANY.RUN SDK](https://github.com/anyrun/anyrun-sdk).
We were also able to improve the quality of the data we provide, based on our actual API data and OpenCTI capabilities.

In this PR we also provide improved documentation and a new source of enrichment - connector for the ANY.RUN TI Lookup

Closes #6121

helene-nguyen reviewed (dismissed):
Waiting for the changes above

helene-nguyen commented:
@semayellow Sorry but the PR cannot be accepted as it is as it brings too much commits at this time, could you handle this ?

semayellow commented:
Hi @helene-nguyen 
It's done. Please check.

helene-nguyen commented:
@semayellow I'm checking it, thanks :)

helene-nguyen commented on external-import/anyrun-feed/docker-compose.yml:
```diff
@@ -1,14 +1,22 @@
 services:
   connector-anyrun-feed:
-    image: opencti/connector-anyrun-feed:latest
+    image: opencti-connector-anyrun-feed:latest
```
Please revert this change

```suggestion
    image: opencti/connector-anyrun-feed:latest
```

helene-nguyen commented on external-import/anyrun-feed/src/anyrun_feed.py:
```diff
@@ -1,56 +1,153 @@
-    def get_feed(self):
-        response = requests.get(
-            self.ti_url, headers={"Authorization": "API-Key {}".format(self.token)}
+        self._organization = self._helper.api.identity.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform.

helene-nguyen commented on external-import/anyrun-feed/src/anyrun_feed.py:
```diff
@@ -1,56 +1,153 @@
+                    externalReferences=self._add_external_references(
+                        feed.get("external_references")
+                    ),
                 )
```
Same here, never use the direct API call when creating Indicators, and create STIX object instead.

helene-nguyen commented on external-import/anyrun-feed/src/anyrun_feed.py:
```diff
@@ -1,56 +1,153 @@
+            refs_identifiers = []
+
+            for ref_info in refs[:15]:
+                external_reference = self._helper.api.external_reference.create(
+                    source_name=ref_info.get("source_name"), url=ref_info.get("url")
+                )
+
+                refs_identifiers.append(external_reference.get("id"))
```
Same here, use STIX2 object for creating external references

helene-nguyen reviewed (commented):
@semayellow Here are some important changes, sorry again for the delay

When writing an OpenCTI connector or any integration that feeds data into the platform, you have two options: build STIX objects and push them through the connector helper using `send_stix2_bunde`, or bypass that and hit the GraphQL API directly to create entities one by one like here. 

Both technically work, but STIX object creation is the recommended choice, for several reinforcing reasons.
STIX bundles flow through the worker pipeline, so ingested object counts, processing rates, and errors are visible in the connector and worker metrics out of the box; direct API calls bypass this instrumentation and leave you blind to how much data is actually going in.

Direct API calls work, but they put you on the wrong side of every tradeoff that matters (data consistency, deduplication, resilience, performance, forward compatibility, and portability).

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+        self._helper = OpenCTIConnectorHelper({})
+        self._config = config
+
+        self._organization = self._helper.api.identity.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+        for ipgeo in destination_ips:
+            if self._opencti_entity["entity_type"] == "IPv4-Addr":
+                opencti_locations = self._helper.api.location.list(search=ipgeo)
+                self._helper.api.stix_core_relationship.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+        """
+        for task in source_tasks:
+            if task.get("threatLevel") == 2:
+                external_reference_task = self._helper.api.external_reference.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+        if indicators:
+            for ioc in indicators:
+                if ioc["threatLevel"] == 0:
+                    new_observable = self._helper.api.stix_cyber_observable.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+                        externalReferences=[external_reference_id],
+                    )
+                    if new_observable["id"] != self._opencti_entity["id"]:
+                        self._helper.api.stix_core_relationship.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+                            description="Detected by ANY.RUN TI Lookup",
+                        )
+                elif ioc["threatLevel"] in (1, 2):
+                    indicator = self._helper.api.indicator.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+                        externalReferences=[external_reference_id],
+                    )
+
+                    self._helper.api.stix_core_relationship.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+        elif anyrun_score == 2:
+            score = 100
+
+        self._helper.api.stix_cyber_observable.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+        )
+
+        # Create and add note to the main observable
+        note = self._helper.api.note.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,388 @@
+        )
+
+        # Create external reference for the child entities
+        external_reference_id = self._helper.api.external_reference.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,239 @@
+
+        self._opencti_entity: dict | None = None
+
+        self._organization = self._helper.api.identity.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,239 @@
+        :return: External reference ID
+        """
+        url = f"https://app.any.run/tasks/{task_uuid}"
+        external_reference_id = self._helper.api.external_reference.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,239 @@
+        :param tags: ANY.RUN task tags
+        """
+        for tag in tags:
+            label = self._helper.api.label.create(value=tag["tag"])
+            self._helper.api.stix_cyber_observable.add_label(
+                id=self._opencti_entity["id"], label_id=label["id"]
+            )
```
Add the label and update the STIX object

helene-nguyen commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,239 @@
+                        externalReferences=[external_reference_id],
+                    )
+                    if new_observable["id"] != self._opencti_entity["id"]:
+                        self._helper.api.stix_core_relationship.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,239 @@
+                        )
+
+                elif ioc["reputation"] in (1, 2):
+                    indicator = self._helper.api.indicator.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

helene-nguyen commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,239 @@
+                        externalReferences=[external_reference_id],
+                    )
+
+                    self._helper.api.stix_core_relationship.create(
```
Please create STIX object and and use send_stix2_bundle to push the bundle into the platform here too

semayellow commented:
Hi @helene-nguyen 
I was able to migrate the generation of indicator, note, external_reference, and relationship objects to the stix2 bundles. However, I'm having trouble creating Observable objects using the same method.

Can u please tell me how I can implement this logic using stix2:

### 1. In case of creating a new Observable
```py
self._helper.api.stix_cyber_observable.create(
    createIndicator=False,
    createdBy=self._identity.get('id'),
    update=True,
    observableData={
            'type': 'IPv4-Addr',
            'x_opencti_description': 'Detected by ANY.RUN TI Lookup',
            'value': '1.1.1.1'
        },
    externalReferences=[external_references]
)
```
### 2. In case of updating an existing Observable
For example, if I need to update the author, score, labels, description and other parameters.

```py
self._helper.api.label.create(value=tag)
self._helper.api.stix_cyber_observable.add_label(
    id=self._opencti_entity['id'],
    label_id=label['id']
)

 self._helper.api.stix_cyber_observable.update_field(
    id=self._opencti_entity['id'],
    input={
        'key': 'x_opencti_score',
        'value': str(anyrun_score)
    }
)
```

semayellow commented:
> Hi @helene-nguyen I was able to migrate the generation of indicator, note, external_reference, and relationship objects to the stix2 bundles. However, I'm having trouble creating Observable objects using the same method.
> 
> Can u please tell me how I can implement this logic using stix2:
> 
> ### 1. In case of creating a new Observable
> ```python
> self._helper.api.stix_cyber_observable.create(
>     createIndicator=False,
>     createdBy=self._identity.get('id'),
>     update=True,
>     observableData={
>             'type': 'IPv4-Addr',
>             'x_opencti_description': 'Detected by ANY.RUN TI Lookup',
>             'value': '1.1.1.1'
>         },
>     externalReferences=[external_references]
> )
> ```
> 
> ### 2. In case of updating an existing Observable
> For example, if I need to update the author, score, labels, description and other parameters.
> 
> ```python
> self._helper.api.label.create(value=tag)
> self._helper.api.stix_cyber_observable.add_label(
>     id=self._opencti_entity['id'],
>     label_id=label['id']
> )
> 
>  self._helper.api.stix_cyber_observable.update_field(
>     id=self._opencti_entity['id'],
>     input={
>         'key': 'x_opencti_score',
>         'value': str(anyrun_score)
>     }
> )
> ```

Solved

semayellow commented:
Hi @helene-nguyen 
Is there anything required from my side to continue the review process?

SamuelHassine committed:
[anyrun] Migrate feed and task connectors to anyrun-sdk and add a TI Lookup connector

This PR refactors the existing ANY.RUN connectors around the official
``anyrun-sdk`` package, and introduces a new internal-enrichment
connector that calls the ANY.RUN Threat Intelligence Lookup API.

Three connectors are touched:

* ``external-import/anyrun-feed``
  - Rewrite the main loop around the ``anyrun-sdk`` ``FeedsConnector`` /
    ``FeedsIterator`` helpers.
  - Initiate a **single** OpenCTI work per run and reuse it for every
    bundle (previously a new work was created for every indicator and
    for the final ``to_processed`` call).
  - Send the identity bundle inside the run instead of the constructor.
  - Forward the parsed ``config.yml`` dict to ``OpenCTIConnectorHelper``
    so that a manual deployment can rely on ``config.yml`` for OpenCTI
    connectivity, as documented in the README.
  - Read the API key from ``anyrun.api_key`` (matching the
    ``config.yml.sample`` / ``ANYRUN_API_KEY`` env var) instead of
    ``anyrun.token``.
  - Load ``CONNECTOR_UPDATE_EXISTING_DATA`` as a boolean instead of an
    integer.
  - Regenerate ``__metadata__/CONNECTOR_CONFIG_DOC.md`` and
    ``connector_config_schema.json`` so the schema actually describes
    the EXTERNAL_IMPORT connector (the previous schema was
    copy/pasted from ``anyrun-lookup``).

* ``internal-enrichment/anyrun-task``
  - Refactor into ``anyrun_sandbox.py``, ``opencti_client.py`` and
    ``config.py`` modules.
  - Replace the buggy ``f"ANYRUN_{name.upper()}"`` env var generator by
    an explicit param-name -> (env var, yaml path) mapping so the env
    var names actually match the documented
    ``ANYRUN_TOR``/``ANYRUN_GEO``/``ANYRUN_MITM``/...
    settings (the previous code was looking for the longer
    ``ANYRUN_OPT_NETWORK_*`` names that no docker-compose / schema
    actually exposes).
  - Build the SHA-256 indicator pattern with the canonical
    ``[file:hashes.'SHA-256' = '<hash>']`` syntax instead of the
    invalid ``[sha256:value = '<hash>']``.
  - Move ``anyrun_env.os_type`` to the top of ``config.yml.sample``
    so the connector finds it whichever OS-specific section is filled.
  - Fix ``ANYRUN_ENV_BITNESS`` (``32``/``64``, not ``32``/``65``) and
    ``ANYRUN_ENV_TYPE`` default (``complete``, not ``complrtr``) in
    ``__metadata__/CONNECTOR_CONFIG_DOC.md``.
  - Drop the obsolete ``main.py`` / ``anyrun_task/`` package and the
    accompanying tests (they targeted the previous structure and would
    not run against the refactor).
  - Read the API key from ``anyrun.api_key``.

* ``internal-enrichment/anyrun-lookup`` (new)
  - Implement a TI Lookup enrichment connector using the
    ``anyrun-sdk`` ``LookupConnector``.
  - Build file-hash indicator patterns with the canonical
    ``[file:hashes.'SHA-256' = '<hash>']`` /
    ``[file:hashes.'SHA-1' = '<hash>']`` / ``[file:hashes.'MD5' = '...']``
    syntax.
  - Forward the parsed ``config.yml`` dict to
    ``OpenCTIConnectorHelper`` so the documented ``config.yml`` flow
    actually works for manual deployment.
  - Read the API key from ``anyrun.api_key``.
  - Ship a ``__metadata__/connector_manifest.json`` describing the new
    connector.

* All three READMEs - document ``CONNECTOR_CONFIDENCE_LEVEL`` as a
  number between 0 and 100 (the OpenCTI confidence range) instead of
  the previous incorrect 1-4 range.

Validated locally:
- isort==6.0.1 --profile black (clean)
- black==25.1.0 --check (clean)
- flake8 --ignore=E,W (clean)
- pylint --disable=all --enable=no_generated_id_stix,no-value-for-parameter,unused-import
  with the linter_stix_id_generator plugin -> 10/10

Closes #6121

SamuelHassine commented on internal-enrichment/anyrun-task/src/config.py:
```diff
@@ -0,0 +1,235 @@
+                f"ANYRUN_{param_name.upper()}",
+                [f"anyrun_{param_name[:3]}", param_name],
+                config,
+            )
```
Fixed: `config.py` now uses an explicit `param_name -> (env var, yaml path)` mapping (`_BASE_PARAMS`) so the env var names match the documented `ANYRUN_TOR` / `ANYRUN_GEO` / `ANYRUN_MITM` / `ANYRUN_RESIDENTIAL_PROXY` / `ANYRUN_RESIDENTIAL_PROXY_GEO` / `ANYRUN_PRIVACY_TYPE` settings exposed in `docker-compose.yml` and `__metadata__/connector_config_schema.json`.

SamuelHassine commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,337 @@
+        """
+        labels = [tag.get("tag") for tag in analysis_summary["analysis"]["tags"]]
+
+        if self._opencti_entity.get("entity_type") == "Url":
+            observable = stix2.URL(
+                value=self._opencti_entity.get("value"),
+                custom_properties={
+                    "x_opencti_labels": labels,
+                    "x_opencti_score": self._get_score(
+                        analysis_summary["analysis"]["scores"]["verdict"]["score"]
+                    ),
+                    "x_opencti_description": "Detected by ANY.RUN Sandbox",
+                    "x_opencti_created_by_ref": self._identity.get("id"),
+                    "x_opencti_external_references": self._get_external_reference(
+                        task_uuid
+                    ),
+                },
+            )
+        else:
+            observable = stix2.File(
+                hashes={
+                    "MD5": analysis_summary["analysis"]["content"]["mainObject"][
```
STIX 2.1 Cyber Observables (`stix2.File` / `stix2.URL` / `stix2.DomainName` / `stix2.IPv4Address`) already get deterministic UUIDv5 ids generated by the `stix2` library from their value-contributing properties (`value`/`hashes`), so no duplicates are created on re-runs. The repo's STIX id linter (`no_generated_id_stix` in `shared/pylint_plugins/check_stix_plugin`) only enforces explicit `id=` for SDOs and Relationships - which we already do - and the connector now passes that pylint check at 10/10.

SamuelHassine commented on internal-enrichment/anyrun-task/src/opencti_client.py:
```diff
@@ -0,0 +1,337 @@
+                    pattern = "[{}:value = '{}']".format(
+                        ANYRUN_INDICATOR_TO_STIX.get(ioc.get("type")),
+                        ioc.get("ioc"),
+                    )
```
Fixed: file-hash IOC patterns now use the canonical `[file:hashes.'SHA-256' = '<hash>']` syntax instead of `[sha256:value = '<hash>']` via a new `ANYRUN_INDICATOR_PATTERNS` template table.

SamuelHassine commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,494 @@
+        pattern = "[{}:value = '{}']".format(
+            stix_algorithm,
+            value,
+        )
```
Fixed: `_create_file_indicator()` now builds the pattern as `[file:hashes.'SHA-256' = '<hash>']` / `[file:hashes.'SHA-1' = '<hash>']` / `[file:hashes.'MD5' = '<hash>']` via the new `HASH_ALGORITHM_TO_STIX` mapping.

SamuelHassine commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,494 @@
+            observable = stix2.IPv4Address(
+                value=self._opencti_entity.get("value"),
+                custom_properties=self._get_custom_properties(lookup_search_results),
+            )
```
Same answer as on `opencti_client.py:120`: STIX 2.1 Cyber Observables (`stix2.File` / `stix2.DomainName` / `stix2.URL` / `stix2.IPv4Address`) already get deterministic UUIDv5 ids generated by the `stix2` library from their value-contributing properties, and the repo's STIX id linter only checks SDOs / Relationships. The connector passes `pylint --enable=no_generated_id_stix,no-value-for-parameter,unused-import` at 10/10.

SamuelHassine commented on external-import/anyrun-feed/src/anyrun_feed.py:
```diff
@@ -1,56 +1,180 @@
+                    ),
+                    update=self._config.update_existing_data,
+                    work_id=self._get_work_id(),
+                )
```
Fixed: `_run_once()` now initiates a **single** work_id per cycle and reuses it for the identity bundle, every chunk of indicators, and the final `to_processed()` call. Failures are propagated by marking the work as `in_error`.

SamuelHassine commented on external-import/anyrun-feed/src/config.py:
```diff
@@ -0,0 +1,51 @@
+    def __init__(self) -> None:
+        self._create_base_config()
+        self.anyrun_api_key = (
+            f"{get_config_variable('ANYRUN_API_KEY', ['anyrun', 'token'], config)}"
```
Fixed: code now reads the API key from `['anyrun', 'api_key']` to match `config.yml.sample` (which already used `anyrun.api_key`) and the `ANYRUN_API_KEY` env var documented in `docker-compose.yml`.

SamuelHassine commented on external-import/anyrun-feed/src/config.py:
```diff
@@ -0,0 +1,51 @@
+            "CONNECTOR_UPDATE_EXISTING_DATA",
+            ["connector", "update_existing_data"],
+            config,
+            isNumber=True,
```
Fixed: `update_existing_data` is now loaded as a boolean (with `default=False`) instead of a number.

SamuelHassine commented on external-import/anyrun-feed/__metadata__/connector_config_schema.json:
```diff
@@ -0,0 +1,78 @@
+      "minLength": 1,
+      "type": "string"
+    },
+    "OPENCTI_TOKEN": {
+      "description": "The API token to connect to OpenCTI.",
+      "type": "string"
+    },
+    "CONNECTOR_NAME": {
+      "default": "ANY.RUN task",
+      "description": "The name of the connector.",
+      "type": "string"
+    },
+    "CONNECTOR_SCOPE": {
+      "default": [
+        "Artifact",
+        "Url"
+      ],
+      "description": "The scope of the connector.",
+      "items": {
+        "type": "string"
+      },
+      "type": "array"
+    },
+    "CONNECTOR_LOG_LEVEL": {
+      "default": "error",
+      "description": "The minimum level of logs to display.",
+      "enum": [
+        "debug",
+        "info",
+        "warn",
+        "warning",
+        "error"
+      ],
+      "type": "string"
+    },
+    "CONNECTOR_TYPE": {
+      "const": "INTERNAL_ENRICHMENT",
+      "default": "INTERNAL_ENRICHMENT",
+      "type": "string"
+    },
```
Fixed: regenerated the schema with the correct `\` (`anyrun-feed_config.schema.json`), `CONNECTOR_TYPE=EXTERNAL_IMPORT`, `CONNECTOR_NAME=ANY.RUN TI Feed` and `CONNECTOR_SCOPE=[stix2]` defaults, plus the actual feed parameters (`ANYRUN_FEED_FETCH_INTERVAL` / `ANYRUN_FEED_FETCH_DEPTH`). `CONNECTOR_CONFIG_DOC.md` is regenerated accordingly.

SamuelHassine commented on internal-enrichment/anyrun-task/__metadata__/CONNECTOR_CONFIG_DOC.md:
```diff
@@ -4,30 +4,28 @@ Below is an exhaustive enumeration of all configurable parameters available, eac
+| ANYRUN_OS_TYPE | `string` | | `windows` `linux` `android`                                                                               |  | `windows`       | Operating system.                                                                                                              |
+| ANYRUN_ENV_VERSION | `string` | | `7` `10` `11`                                                                                             |  | `10`            | Windows environment version.                                                                                                   |
+| ANYRUN_ENV_BITNESS | `string` | | `32` `65`                                                                                                 |  | `64`            | Operating system bitness: 32 or 64 for Windows, 64 for Linux and Android.                                                      |
+| ANYRUN_ENV_TYPE | `string` | | `complete` `development`                                                                                  |  | `complrtr`      | Windows environment type. You can select development env for OS Windows 10 x64. For all other cases, complete env is required. |
```
Fixed: `ANYRUN_ENV_BITNESS` possible values are now `32 / 64` (no more bogus `65`) and `ANYRUN_ENV_TYPE` default is now `complete` (the typo `complrtr` is gone).

SamuelHassine commented on internal-enrichment/anyrun-task/README.md:
```diff
@@ -13,71 +21,130 @@ The ANY.RUN Task connector analyzes URL and Artifact observables in the ANY.RUN
+| `connector_name`             | `CONNECTOR_NAME`                 | Yes       | A connector name to be shown in OpenCTI.                                                                                                                                                     |
+| `connector_scope`            | `CONNECTOR_SCOPE`                | Yes       | Supported scope. E. g., `text/html`.                                                                                                                                                         |                     
+| `connector_auto`             | `CONNECTOR_AUTO`                 | Yes       | Enable/disable auto-enrichment of observables.                                                                                                                                               |
+| `connector_confidence_level` | `CONNECTOR_CONFIDENCE_LEVEL`     | Yes       | The default confidence level for created sightings (a number between 1 and 4).                                                                                                               |
```
Fixed in all three READMEs (anyrun-task, anyrun-lookup, anyrun-feed): `CONNECTOR_CONFIDENCE_LEVEL` is now documented as `a number between 0 and 100, where 0 = Unknown and 100 = Fully trusted`.

SamuelHassine commented on external-import/anyrun-feed/README.md:
```diff
@@ -13,58 +21,86 @@ The ANY.RUN connector imports threat intelligence feed data from the ANY.RUN Int
+| `connector_name`             | `CONNECTOR_NAME`                 | Yes       | A connector name to be shown in OpenCTI.                                                                                                                                                     |
+| `connector_scope`            | `CONNECTOR_SCOPE`                | Yes       | Supported scope. E. g., `text/html`.                                                                                                                                                         |                     
+| `connector_auto`             | `CONNECTOR_AUTO`                 | Yes       | Enable/disable auto-enrichment of observables.                                                                                                                                               |
+| `connector_confidence_level` | `CONNECTOR_CONFIDENCE_LEVEL`     | Yes       | The default confidence level for created sightings (a number between 1 and 4).                                                                                                               |
```
Same fix as on `anyrun-task/README.md`: confidence_level is now documented as `a number between 0 and 100, where 0 = Unknown and 100 = Fully trusted`.

SamuelHassine commented on internal-enrichment/anyrun-lookup/README.md:
```diff
@@ -0,0 +1,285 @@
+| `connector_name`             | `CONNECTOR_NAME`                 | Yes       | A connector name to be shown in OpenCTI.                                                                                                                                                     |
+| `connector_scope`            | `CONNECTOR_SCOPE`                | Yes       | Supported scope. E. g., `text/html`.                                                                                                                                                         |                     
+| `connector_auto`             | `CONNECTOR_AUTO`                 | Yes       | Enable/disable auto-enrichment of observables.                                                                                                                                               |
+| `connector_confidence_level` | `CONNECTOR_CONFIDENCE_LEVEL`     | Yes       | The default confidence level for created sightings (a number between 1 and 4).                                                                                                               |
```
Same fix as on the other READMEs: confidence_level is now documented as `a number between 0 and 100, where 0 = Unknown and 100 = Fully trusted`.

SamuelHassine commented on external-import/anyrun-feed/src/anyrun_feed.py:
```diff
@@ -1,56 +1,180 @@
+        self._config = config
+        self._helper = OpenCTIConnectorHelper({})
+
+        self._identity = stix2.Identity(
```
Fixed: `OpenCTIConnectorHelper` is now instantiated with the parsed `config` dict from `config.py`, so a manual deployment with `config.yml` actually picks up the OpenCTI / RabbitMQ settings documented in `config.yml.sample`.

SamuelHassine commented on internal-enrichment/anyrun-lookup/src/anyrun_lookup.py:
```diff
@@ -0,0 +1,494 @@
+
+class AnyRunTILookup:
+    def __init__(self, config: Config):
+        self._helper = OpenCTIConnectorHelper({})
```
Same fix as on `anyrun-feed`: `OpenCTIConnectorHelper` is now instantiated with the parsed `config` dict, so the `config.yml` flow described in the README actually works for manual deployments.

SamuelHassine reviewed (approved):
Full review done.

Rebased the entire change set onto the current master and squashed everything into a single GPG-signed commit (the original 5-commit history was carrying merge conflicts with master and stale fix-up commits). Every one of the 15 Copilot review threads has been addressed and resolved:

Code fixes:
- anyrun-task `config.py`: replaced the buggy `f"ANYRUN_{name.upper()}"` env var generator by an explicit param-name -> (env var, yaml path) mapping so the env var names actually match `ANYRUN_TOR` / `ANYRUN_GEO` / `ANYRUN_MITM` / `ANYRUN_RESIDENTIAL_PROXY` / `ANYRUN_RESIDENTIAL_PROXY_GEO` / `ANYRUN_PRIVACY_TYPE` as documented in `docker-compose.yml` and `__metadata__/connector_config_schema.json`.
- anyrun-task `opencti_client.py` and anyrun-lookup `anyrun_lookup.py`: file-hash IOC patterns now use the canonical `[file:hashes.'SHA-256' = '<hash>']` syntax (plus SHA-1 / MD5 for the lookup connector) instead of the invalid `[sha256:value = '<hash>']`.
- anyrun-feed `anyrun_feed.py`: refactored the main loop so a **single** OpenCTI work is initiated per cycle and reused for the identity bundle, every chunk of indicators and the final `to_processed()` call (previously a new work was created for every indicator).
- anyrun-feed `config.py`: read the API key from `anyrun.api_key` (matching `config.yml.sample` and `ANYRUN_API_KEY`), and load `CONNECTOR_UPDATE_EXISTING_DATA` as a boolean (not an integer).
- anyrun-feed and anyrun-lookup __main__: instantiate `OpenCTIConnectorHelper` with the parsed `config.yml` dict so a manual deployment actually picks up the OpenCTI / RabbitMQ settings documented in `config.yml.sample`.

Schema and docs:
- anyrun-feed `__metadata__/connector_config_schema.json` and `CONNECTOR_CONFIG_DOC.md` regenerated with the correct `$id`, `CONNECTOR_TYPE=EXTERNAL_IMPORT`, name / scope and parameters (the previous schema was a copy of anyrun-lookup).
- anyrun-task `__metadata__/CONNECTOR_CONFIG_DOC.md`: fixed `ANYRUN_ENV_BITNESS` possible values (`32` / `64`, not `32` / `65`) and `ANYRUN_ENV_TYPE` default (`complete`, not `complrtr`).
- All three READMEs (anyrun-task, anyrun-lookup, anyrun-feed): `CONNECTOR_CONFIDENCE_LEVEL` is now documented as a number between 0 and 100 instead of 1-4.

About the two threads claiming STIX Cyber Observables require explicit `id=`: STIX 2.1 SCOs already get deterministic UUIDv5 ids generated by the `stix2` library from their value-contributing properties, so no duplicates are created on re-runs. The repo's `no_generated_id_stix` linter in `shared/pylint_plugins/check_stix_plugin` only enforces explicit `id=` for SDOs and Relationships (which we already do everywhere), and the three connectors now pass that pylint check at 10/10. Resolved as not applicable.

PR title also realigned with the master `[<slug>] <description>` convention. CI is fully green.
